File: reference-impls/typescript/src/serializer.ts
import { validate } from './validator.js'
import type { UACPDocument } from './types.js'

export const UACP_VERSION = '0.7.0'

export function parse(json: string | object): UACPDocument {
  const doc = typeof json === 'string' ? JSON.parse(json) : json
  const result = validate(doc)
  if (!result.ok) {
    throw new Error(`UACP parse failed:\n${result.errors!.join('\n')}`)
  }
  return doc as UACPDocument
}

export function serialize(doc: UACPDocument): string {
  const result = validate(doc)
  if (!result.ok) {
    throw new Error(`UACP serialize failed: document is not valid:\n${result.errors!.join('\n')}`)
  }
  return JSON.stringify(doc)
}

File: reference-impls/python/uacp/serializer.py
from __future__ import annotations
import json
from .types import UACPDocument
from .validator import validate


def parse(data: str | dict) -> UACPDocument:
    if isinstance(data, str):
        try:
            raw = json.loads(data)
        except json.JSONDecodeError as e:
            raise ValueError(f'UACP parse failed: invalid JSON: {e}') from e
    else:
        raw = data
    result = validate(raw)
    if not result['ok']:
        raise ValueError('UACP parse failed:\n' + '\n'.join(result['errors']))
    return UACPDocument.model_validate(raw)


def serialize(doc: UACPDocument) -> str:
    raw = doc.model_dump(exclude_none=True)
    result = validate(raw)
    if not result['ok']:
        raise ValueError('UACP serialize failed: document is not valid:\n' + '\n'.join(result['errors']))
    return json.dumps(raw)
